README.md
Mr. Provisioner
===============

Provision a host using [Mr. Provisioner](https://github.com/mr-provisioner/mr-provisioner).

This role will upload a kernel and initrd file to Mr. Provisioner, then
configure the machine (based on inventory_hostname) to use the given kernel,
initrd, and preseed file. Then, it will do a PXE reboot and wait for the host
to come online.

Role Variables
--------------

``mr_provisioner_do_provision``: Defaults to True. Disable to skip all role
tasks. This is useful if all you want is the modules.

If mr_provisioner_do_provision is set, all of the following variables are
required:
- ``mr_provisioner_machine_name``: Machine's name.
- ``mr_provisioner_kernel_description``: Unique kernel description to use in
  provisioner.
- ``mr_provisioner_initrd_description``: Unique initrd description to use in
  provisioner.
- ``mr_provisioner_kernel_path``: Local path to kernel file for uploading to
  provisioner.
- ``mr_provisioner_initrd_path``: Local path to initrd file for uploading to
  provisioner.
- ``mr_provisioner_url``: Mr. Provisioner URL in the form of i.e.
  'http://192.168.0.3:5000'
- ``mr_provisioner_auth_token``: Auth token from Mr. Provisioner.
- ``mr_provisioner_arch``: Image architecture.
- ``mr_provisioner_subarch``: Machine subarchitecture.
- ``machine_user``: Machine SSH user that wait_for_connection should try authenticating as.

Usage
-----

    - hosts: provision_this_machine
      vars:
	mr_provisioner_machine_name: "if you don't assign it, it will use inventory hostname"
        mr_provisioner_kernel_description: "debian-installer staging build 495"
        mr_provisioner_initrd_description: "debian-installer staging build 495"
        mr_provisioner_kernel_path: "./builds/debian-staging/495/linux"
        mr_provisioner_initrd_path: "./builds/debian-staging/495/initrd.gz"
        mr_provisioner_url: "http://192.168.0.3:5000"
        mr_provisioner_auth_token: "MYSUPERFANCYTOKENFROMPROVISIONER"
        mr_provisioner_preseed_name: "danrue/erp-17.08-debian-20170727.567664d4"
        mr_provisioner_preseed_path: "./preseeds/erp-17.08-generic"
        mr_provisioner_arch: "arm64"
        mr_provisioner_subarch: "efi"
	machine_user: "myspecialuser"
      roles:
        - role: Linaro.mr-provisioner

    - hosts: target
      gather_facts: no
      tasks:
		- name: Wait for host for 3600 seconds, but only start checking after 60.
		  wait_for_connection:
			delay: 60
			timeout: 3600

Role Modules
------------

This role contains four ansible modules:
- ``mr_provisioner_image``: Handles uploading image files to Mr. Provisioner.
- ``mr_provisioner_machine_provision``: Handles provisioning a host in Mr.
  Provisioner.
- ``mr_provisioner_preseed``: Handles uploading preseed files to Mr. Provisioner.
- ``mr_provisioner_get_ip``: Handles fetching the provisioned machine's IP from Mr. Provisioner.

By default, these modules are used by the tasks in the role. They may also be
used outside the role if the included role tasks are not suitable.

    - hosts: all
      roles:
        - role: Linaro.mr-provisioner
          mr_provisioner_do_provision: False

By setting mr_provisioner_do_provision to False, the modules will be made
available but no tasks will run.

Caveats
-------

Most of these behaviors can be fixed but in the meantime, FYI!

- Use unique image descriptions. The kernel and initrd files are only uploaded
  once. If they're changed locally but the description is the same, they will
  not be re-uploaded. If it finds someone else's image with the same
  description, it will use it. If there are multiple images with the same
  description, it will use the first one found (non deterministically).

See Also
--------

Role to retrieve a kernel and initrd file from Linaro's ERP:
[erp-get-build](https://galaxy.ansible.com/Linaro/erp-get-build/)

License
-------

BSD

Author Information
------------------

Dan Rue <dan.rue@linaro.org>


CHANGELOG.md
# CHANGELOG

## v1.0.4 (2018-04-03)

Feature:

  - Add support for upload preseed files.

## v1.0.3 (2018-02-08)

Improvement:
  - Increase `wait_for_connection` timeout to adapt slow network env.

## v1.0.2 (2018-01-19)

Feature:

  - Add support for multiple architectures

Dependency:

  - [Mr. Provisioner](https://github.com/Linaro/mr-provisioner) >= v0.2.8


library/test.py
import requests
#from urlparse import urljoin
from future.standard_library import install_aliases
install_aliases()

from urllib.parse import urljoin

# curl -X POST "http://172.27.80.1:5000/api/v1/image"
# -H "accept: application/json"
# -H "Authorization: DEADBEEF"
# -H "content-type: multipart/form-data"
# -F "file=@linux;type="
# -F "q={ "description": "Example image",
#         "type": "Kernel",
#         "arch": "arm64",
#         "public": false,
#         "known_good": true } "
headers = {'Authorization': '/T9kmICCxhhk0Ec6kCqudgXwwWNTzNrrqmuCTCAwA2U='}
url = urljoin("http://172.27.80.1:5000", "/api/v1/image")
files = {'file': open('../../../builds/debian-staging/476/linux', 'rb')}
data = {'q': '{"name": "drue test 476", "description": "", "type": "Kernel", "arch": "arm64"}'
        }
r = requests.post(url, files=files, data=data, headers=headers)
if r.status_code != 200:
    print('Error posting {}, HTTP {}, {}'.format(url,
                     r.status_code, r.reason))
print(r.status_code)
print(r.json())
import pdb; pdb.set_trace()



library/mr_provisioner_get_ip.py
#!/usr/bin/env python2.7
# -*- coding: utf-8 -*-

import requests
import json
try:
    from urllib.parse import urljoin
except ImportError:
    from urlparse import urljoin
try:
        from urllib import quote  # Python 2.X
except ImportError:
        from urllib.parse import quote  # Python 3+

ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}

DOCUMENTATION = '''
---
module: mr_provisioner_get_ip

short_description: This is a short module fetching the ip of the machine being
provisioned from MrP. It should be then used to create a new host (dynamically)
and then do the wait_for_connection on that new host. (sadly to do that one
needs to go all the way back to the playbook).

version_added: "1.0"

description:
    - "This module has been designed to compliment the provisioning role,
    making it able to fetch the IP from MrP, making use of the available API
    for it. It should be noted that the current behaviour is of fetching the
    reserved address for an dynamic-reserved interface. If it's static or
    dynamic, the module will fetch the last address given by KEA"

options:
    mrp_url:
        description:
            - This is the URL of the Mr Provisioner
        required: true
    mrp_token:
        description:
            - This is the authentication token for MrP's API
        required: true
    machine_name:
        description:
            - This is the machine name as shown in MrP
        required: true
    interface_name:
        description:
            - This is the name of the machine's interface you'd like the IP of.
        required: true

author:
    - Baptiste Gerondeau (baptiste.gerondeau@linaro.org)
'''

EXAMPLES = '''
- name: Get IP of provisioned machine
  mr_provisioner_get_ip:
    mrp_url: "{{ mr_provisioner_url }}"
    mrp_token: "{{ mr_provisioner_auth_token }}"
    machine_name: "{{ mr_provisioner_machine_name }}"
    interface_name: "{{ mr_provisioner_interface_name|default('eth1') }}"
  register: get_ip
- debug: var=get_ip

#Note that you get the ip fetched via get_ip['ip'] : you NEED to register get_ip
#This seems overly complicated but I've found no other way to do it...
'''

RETURN = '''
ip:
    description: An.... IP !! (v4 because MrP doesn't do v6)
    type: str
'''


from ansible.module_utils.basic import AnsibleModule

class ProvisionerError(Exception):
    def __init__(self, message):
        super(ProvisionerError, self).__init__(message)

class IPGetter(object):
    def __init__(self, mrpurl, mrptoken, machine_id, interface_name =
                 'eth1'):
        self.mrp_url = mrpurl
        self.mrp_token = mrptoken
        self.interface = interface_name
        self.machine_id = machine_id
        self.machine_ip = ''

    def get_interfaces(self):
        headers = {'Authorization': self.mrp_token}
        url = urljoin(self.mrp_url,
                      "/api/v1/machine/{}/interface".format(self.machine_id))
        r = requests.get(url, headers=headers)
        if r.status_code != 200:
            raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(self.mrp_url, r.status_code,
                                                         r.reason))
        if len(r.json()) == 0:
            raise ProvisionerError('Error no machine with id "{}"'.format(self.machine_id))

        return r.json()

    def get_ip(self):
        try:
            interfaces = self.get_interfaces()
        except ProvisionerError as e:
            print('Could not fetch interface for machine : "{}"'.format(e))
            return 'FAILURE'

        for i in interfaces:
            if str(i['identifier']) == self.interface:
                if str(i['config_type_v4']) == 'dynamic-reserved' and str(i['configured_ipv4']):
                    return i['configured_ipv4']
                else:
                    return i['lease_ipv4']

def get_machine_by_name(mrp_token, mrp_url, machine_name):
    """ Look up machine by name """
    headers = {'Authorization': mrp_token}
    q = '(= name "{}")'.format(quote(machine_name))
    url = urljoin(mrp_url, "/api/v1/machine?q={}&show_all=false".format(q))
    r = requests.get(url, headers=headers)
    if r.status_code != 200:
        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(mrp_url,
                         r.status_code, r.reason))
    if len(r.json()) == 0:
       raise ProvisionerError('Error no assigned machine found with name "{}"'.
                    format(machine_name))
    if len(r.json()) > 1:
       raise ProvisionerError('Error more than one machine found with name "{}", {}'.
                    format(machine_name, r.json()))
    return r.json()[0]

def run_module():
    module_args = dict(
        mrp_url = dict(type='str', required=True),
        mrp_token = dict(type='str', required=True),
        machine_name = dict(type='str', required=True),
        interface_name = dict(type='str', required=False),
    )

    result = dict(
        changed=False,
        debug={},
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True,
    )

    if module.check_mode:
        return result

    machine_id = get_machine_by_name(module.params['mrp_token'],
                                     module.params['mrp_url'],
                                     module.params['machine_name'])['id']

    if module.params['interface_name']:
        ipgetter = IPGetter(module.params['mrp_url'], module.params['mrp_token'],
                            machine_id,
                            module.params['interface_name'])
    else:
        ipgetter = IPGetter(module.params['mrp_url'], module.params['mrp_token'],
                            machine_id)
    try:
        machine_ip = str(ipgetter.get_ip())
    except ProvisionerError as e:
        module.fail_json(msg='Could not get IP error : "{}"'.format(e),
                         **result)

    if machine_ip:
        result['ip'] = machine_ip
        result['json'] = { 'status': 'ok' }
        result['changed'] = True
    else:
        module.fail_json(msg='Failure to fetch IP from MrP', **result)

    module.exit_json(**result)

def main():
    run_module()

if __name__ == '__main__':
    main()



library/mr_provisioner_preseed.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import json
import requests
try:
    from urllib.parse import urljoin
except ImportError:
    from urlparse import urljoin

ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}

DOCUMENTATION = '''
---
module: mr-provisioner-preseed
short_description: Manage preseed files in Mr. Provisioner
description:
    Implemented:
        - Upload new preseed
        - Discover existing preseeds by a given name.
    Not implemented:
        - modifying existing preseed
        - deleting existing preseed
options:
    name:
        description:
            - Name of the preseed
        required: true
    description:
        description:
            - Description of the preseed
        required: false
    path:
        description: Local file path to preseed file.
        required: true
    url:
        description: url to provisioner instance in the form of http://172.27.80.1:5000/
        required: true
    token:
        description: Mr. Provisioner auth token
        required: true
    known_good:
        description: Mark known good. Default false.
        required: false
    public:
        description: Mark public. Default false.
        required: false
author:
    - Jorge Niedbalski <jorge.niedbalski@linaro.org>
    - Baptiste Gerondeau <baptiste.gerondeau@linaro.org>
'''

EXAMPLES = '''
# Upload a preseed file to a MrProvisioner install.
- name: moonshot-generic-preseed
  path: ./preseeds/moonshot-generic.preseed.txt
  url: http://172.27.80.1:5000/
  token: "{{ provisioner_auth_token }}"

# Uses existing file from MrProvisioner
- name: test_preseed
  path: ''
  url: http://172.0.0.1:5000
  token: "{{ fancy_token }}"
'''

RETURN = '''
  id: auto-assigned preseed id
  description: preseed description
  name: preseed name
  type: user defined type (default: preseed)
  user: User that owns the preseed
  known_good: true/false
  public: true/false
'''

from ansible.module_utils.basic import AnsibleModule

class ProvisionerError(Exception):
    def __init__(self, message):
        super(ProvisionerError, self).__init__(message)

class PreseedUploader(object):
    """ This class handles the job of uploading a preseed file to MrP.
        It shall only be called if there is a file to be uploaded, else you can
        fetch the thing via the regular call in the Ansible role"""
    def __init__(self, mrp_url, mrp_token, preseed_file, preseed_name,
                 preseed_type, preseed_desc='', preseed_knowngood=False,
                 preseed_public=False):
        self.url = mrp_url
        self.authhead = {'Authorization': mrp_token}
        self.file = preseed_file
        self.name = preseed_name
        self.type = preseed_type
        self.id = None
        self.desc = preseed_desc
        self.knowngood = preseed_knowngood
        self.public = preseed_public

    def _check_for_existence(self):
        url = urljoin(self.url, '/api/v1/preseed?show_all=true')
        r = requests.get(url, headers=self.authhead)
        if r.status_code != 200:
            raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url,
                                                r.status_code, r.reason))

        for preseed in r.json():
            if preseed['name'] == self.name:
                self.id = preseed['id']
                return True

        return False

    def _get_preseed_from_file(self):
        json_preseed = {}
        contents = ''

        with open(self.file, 'r') as fd:
            lines = fd.readlines()

        for line in lines:
            contents += line

        json_preseed['content'] = contents
        json_preseed['name'] = self.name
        json_preseed['type'] = self.type
        json_preseed['public'] = self.public
        json_preseed['known_good'] = self.knowngood
        if self.desc != '':
            json_preseed['description'] = self.desc

        return json_preseed

    def upload_preseed(self):
        """Uploads preseed. Should check first that preseed doesn't exist, else
        it modifies preseed (separate function ?). Post to upload, Put to
        modify (+ id). Maybe implement a jinja2 syntax check ? But that should
        be done on mrp's side"""
        res = {}
        try:
            exists = self._check_for_existence()
        except ProvisionerError as e:
            res['error'] = str(e)
            return res

        if not exists and self.file == '':
            res['error'] = 'Preseed does not exist and file not given'
            return res

        if self.id != None and self.file != '':     #Exists and file given
            try:
                res = self._modify_preseed(method='PUT')
            except ProvisionerError as e:
                res['error'] = str(e)
            return res
        elif self.file != '':        #Doesn't exist and file given
            try:
                res = self._modify_preseed(method='POST')
            except ProvisionerError as e:
                res['error'] = str(e)
            return res
        else:       #Exists and file not given, is it useful fetching contents?
            return res

    def _modify_preseed(self, method):
        preseed = self._get_preseed_from_file()
        url = urljoin(self.url, '/api/v1/preseed')
        if method == 'PUT':
            if self.id == None:
                raise ProvisionerError('preseed ID is undefined, please use upload_preseed')
            url_id = '/api/v1/preseed/' + str(self.id)
            url = urljoin(self.url, url_id)
            r = requests.put(url, headers=self.authhead, data=json.dumps(preseed))
            if r.status_code != 200:
                raise ProvisionerError('Error putting preseed {} at ID {}, \
                                   HTTP {} {}'.format(self.name, url, r.status_code, r.reason))
        elif method == 'POST':
            r = requests.post(url, headers=self.authhead, data=json.dumps(preseed))
            if r.status_code != 201:
                raise ProvisionerError('Error posting preseed {}, \
                                       HTTP {} {}'.format(self.name, r.status_code, r.reason))
        else:
            raise ProvisionerError('Bad _modify_preseed call')
        return r.json()

def run_module():
    module_args = dict(
        description=dict(type='str', required=False, default=''),
        name=dict(type='str', required=True),
        type=dict(type='str', required=False, default='preseed'),
        path=dict(type='str', required=False, default=''),
        url=dict(type='str', required=True),
        token=dict(type='str', required=True),
        known_good=dict(type='bool', required=False, default=False),
        public=dict(type='bool', required=False, default=False),
    )

    result = dict(
        changed=False,
        debug={},
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True,
    )

    if module.check_mode:
        return result

    uploader = PreseedUploader(module.params['url'],
    module.params['token'], module.params['path'],
    module.params['name'], module.params['type'],
    module.params['description'], module.params['known_good'],
    module.params['public'])

    try:
        res = uploader.upload_preseed()
    except ProvisionerError as e:
        module.fail_json(msg=str(e), **result)

    if 'error' in res:
        module.fail_json(msg=res['error'], **result)

    result['json'] = res
    result['changed'] = True

    module.exit_json(**result)

def main():
    run_module()

if __name__ == '__main__':
    main()


library/mr_provisioner_machine_provision.py
#!/usr/bin/python

import json
import requests

from future.standard_library import install_aliases
install_aliases()

try:
    from urlparse import urljoin    #Python2
except ImportError:
    from urllib.parse import urljoin    #Python3

try:
    from urllib import quote
except ImportError:
    from urllib.parse import quote

ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}

DOCUMENTATION = '''
---
module: mr-provisioner-machine-provision

short_description: Provision a machine in Mr. Provisioner.

description:
    Implemented:
        - Set machine's initrd, kernel
        - Provision machine
        - Wait for machine to come online
    Not implemented:
        -
        -

options:
    machine_name:
        description: Machine name
        required: true
    kernel_description:
        description: kernel description
        required: true
    initrd_description:
        description: initrd description
        required: true
    arch:
        description: Image architecture. e.g. arm64, x86_64
        required: true
    subarch:
        description: Machine subarchitecture. e.g. efi, bios
        required: true
    preseed_name:
        description: name of preseed to use.
        required: true
    kernel_options:
        description: kernel boot command line
        required: false
    url:
        description: url to provisioner instance in the form of http://192.168.0.3:5000/
        required: true
    token:
        description: Mr. Provisioner auth token
        required: true

author:
    - Dan Rue <dan.rue@linaro.org>
'''

EXAMPLES = '''

'''

RETURN = '''

'''

from ansible.module_utils.basic import AnsibleModule

class ProvisionerError(Exception):
    def __init__(self, message):
        super(ProvisionerError, self).__init__(message)


def machine_provision(url, token, machine_id):
    """ enables netboot on the machine and pxe boots it """
    headers = {'Authorization': token}
    url = urljoin(url, "/api/v1/machine/{}/state".format(machine_id))

    data = json.dumps({'state': 'provision'})

    r = requests.post(url, headers=headers, data=data)

    if r.status_code not in [200, 202]:
        raise ProvisionerError('Error PUTing {}, HTTP {} {}'.format(url,
                         r.status_code, r.reason))
    return r.json()


def set_machine_parameters(url, token, machine_id, initrd_id=None,
                           kernel_id=None, kernel_opts="", preseed_id=None, subarch=None):
    """ Set parameters on machine specified by machine_id """
    headers = {'Authorization': token}
    url = urljoin(url, "/api/v1/machine/{}".format(machine_id))

    parameters = {}
    if initrd_id:
        parameters['initrd_id'] = initrd_id
    if kernel_id:
        parameters['kernel_id'] = kernel_id
    if preseed_id:
        parameters['preseed_id'] = preseed_id
    if subarch:
        parameters['subarch'] = subarch
    if kernel_opts:
        parameters['kernel_opts'] = kernel_opts

    parameters['netboot_enabled'] = True

    data = json.dumps(parameters)

    r = requests.put(url, headers=headers, data=data)

    if r.status_code != 200:
        raise ProvisionerError('Error PUT {}, HTTP {} {}'.format(url,
                         r.status_code, r.reason))
    return r.json()

def get_machine_by_name(url, token, machine_name):
    """ Look up machine by name """
    headers = {'Authorization': token}
    q = '(= name "{}")'.format(quote(machine_name))
    url = urljoin(url, "/api/v1/machine?q={}&show_all=false".format(q))
    r = requests.get(url, headers=headers)
    if r.status_code != 200:
        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url,
                         r.status_code, r.reason))
    if len(r.json()) == 0:
        raise ProvisionerError('Error no assigned machine found with name "{}"'.
                format(machine_name))
    if len(r.json()) > 1:
        raise ProvisionerError('Error more than one machine found with name "{}", {}'.
                format(machine_name, r.json()))
    return r.json()[0]

def get_preseed_by_name(url, token, preseed_name):
    """ Look up preseed by name """
    headers = {'Authorization': token}
    url = urljoin(url, "/api/v1/preseed?show_all=true")
    r = requests.get(url, headers=headers)
    if r.status_code != 200:
        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url,
                         r.status_code, r.reason))
    for preseed in r.json():
        if preseed['name'] == preseed_name:
            del preseed['content'] # we don't need it, and it's really big
            return preseed

    raise ProvisionerError('Error no preseed found with name "{}"'.
            format(preseed_name))

def get_image_by_description(url, token, image_type, description, arch):
    """ Look up image by description """
    headers = {'Authorization': token}
    url = urljoin(url, "/api/v1/image?show_all=true")
    r = requests.get(url, headers=headers)
    if r.status_code != 200:
        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url,
                         r.status_code, r.reason))
    found_image = True
    for image in r.json():
        if (image['description'] == description and
            image['type'] == image_type and
            image['arch'] == arch):
            return image
    msg = "Error finding image of type '{}' and description '{}'".format(
        image_type, description)
    raise ProvisionerError(msg)

def run_module():
    # define the available arguments/parameters that a user can pass to
    # the module
    module_args = dict(
        machine_name=dict(type='str', required=True),
        kernel_description=dict(type='str', required=True),
        initrd_description=dict(type='str', required=True),
        arch=dict(type='str', required=True),
        subarch=dict(type='str', required=True),
        preseed_name=dict(type='str', required=True),
        kernel_options=dict(type='str', required=False),
        url=dict(type='str', required=True),
        token=dict(type='str', required=True),
    )

    result = dict(
        changed=False,
        debug={},
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True
    )

    if module.check_mode:
        return result


    # Look up machine, verify assignment
    try:
        machine = get_machine_by_name(module.params['url'],
                                      module.params['token'],
                                      module.params['machine_name'])
    except ProvisionerError as e:
        module.fail_json(msg=str(e), **result)
    result['debug']['machine'] = machine

    # Look up kernel, initrd
    try:
        kernel_id = get_image_by_description(module.params['url'],
                                         module.params['token'],
                                         "Kernel",
                                         module.params['kernel_description'],
                                         module.params['arch'])
        initrd_id = get_image_by_description(module.params['url'],
                                         module.params['token'],
                                         "Initrd",
                                         module.params['initrd_description'],
                                         module.params['arch'])
    except ProvisionerError as e:
        module.fail_json(msg=str(e), **result)
    result['debug']['kernel_id'] = kernel_id
    result['debug']['initrd_id'] = initrd_id

    # Look up kernel, initrd, and preseed IDs
    try:
        preseed = get_preseed_by_name(module.params['url'],
                                      module.params['token'],
                                      module.params['preseed_name'])
    except ProvisionerError as e:
        module.fail_json(msg=str(e), **result)
    result['debug']['preseed'] = preseed

    # Set kernel, initrd, preseed on machine
    try:
        machine_state = set_machine_parameters(module.params['url'],
                                      module.params['token'],
                                      machine_id=machine['id'],
                                      initrd_id=initrd_id['id'],
                                      kernel_id=kernel_id['id'],
                                      kernel_opts=module.params['kernel_options'],
                                      preseed_id=preseed['id'],
                                      subarch=module.params['subarch'])

    except ProvisionerError as e:
        module.fail_json(msg=str(e), **result)
    result['machine_state'] = machine_state

    # Reboot/provision
    try:
        machine_state = machine_provision(module.params['url'],
                                      module.params['token'],
                                      machine_id=machine['id'])

    except ProvisionerError as e:
        module.fail_json(msg=str(e), **result)
    result['machine_provision'] = machine_state

    module.exit_json(**result)

def main():
    run_module()

if __name__ == '__main__':
    main()


library/mr_provisioner_image.py
#!/usr/bin/python

import json
import requests

from future.standard_library import install_aliases
install_aliases()

try:
    from urlparse import urljoin    #Python2
except ImportError:
    from urllib.parse import urljoin    #Python3

ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}

DOCUMENTATION = '''
---
module: mr-provisioner-image

short_description: Manage machine images in Mr. Provisioner

description:
    Implemented:
        - Upload new image
        - Discover existing images matching a given description.
    Not implemented:
        - modifying existing image (such as known_good/public)
        - deleting existing image

options:
    description:
        description:
            - Name of the image
        required: true
    type:
        description:
            - Image type. May be 'Kernel' or 'Initrd'.
        required: true
    arch:
        description: Image architecture. e.g. arm64, x86_64
        required: true
    path:
        description: Local file path to image file.
        required: true
    url:
        description: url to provisioner instance in the form of http://172.27.80.1:5000/
        required: true
    token:
        description: Mr. Provisioner auth token
        required: true
    known_good:
        description: Mark known good. Default false.
        required: false
    public:
        description: Mark public. Default false.
        required: false

author:
    - Dan Rue <dan.rue@linaro.org>
'''

EXAMPLES = '''
# Upload a linux kernel image
- description: debian-installer staging build 471
  type: Kernel
  arch: arm64
  path: ./builds/staging/427/linux
  url: http://172.27.80.1:5000/
  token: "{{ provisioner_auth_token }}"
'''

RETURN = '''
  id: auto-assigned image id
  description: image description
  name: auto-assigned image name
  type: Kernel or Initrd
  upload_date: Date of upload
  user: User that owns the image
  known_good: true/false
  public: true/false
  arch: arm64
'''

from ansible.module_utils.basic import AnsibleModule

def run_module():
    # define the available arguments/parameters that a user can pass to
    # the module
    module_args = dict(
        description=dict(type='str', required=True),
        type=dict(type='str', required=True),
        arch=dict(type='str', required=True),
        path=dict(type='str', required=True),
        url=dict(type='str', required=True),
        token=dict(type='str', required=True),
        known_good=dict(type='bool', required=False, default=False),
        public=dict(type='bool', required=False, default=False),
    )

    result = dict(
        changed=False
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True
    )

    if module.check_mode:
        return result

    allowed_types = ["Kernel", "Initrd"]
    if module.params['type'] not in allowed_types:
        module.fail_json(msg="error: type is '{}'; must be one of {}".format(
                         module.params['type'], allowed_types), **result)

    # Determine if image is already uploaded
    headers = {'Authorization': module.params['token']}
    url = urljoin(module.params['url'], "/api/v1/image?show_all=true")
    r = requests.get(url, headers=headers)
    if r.status_code != 200:
        module.fail_json(msg='Error fetching {}, HTTP {} {}'.format(url,
                         r.status_code, r.reason), **result)
    for image in r.json():
        if (image['description'] == module.params['description'] and
            image['type'] == module.params['type'] and
            image['arch'] == module.params['arch']):
                #XXX Not implemented: modify existing image
                result['json'] = image
                module.exit_json(**result)

    # Image does not yet exist. Upload it.
    # curl -X POST "http://172.27.80.1:5000/api/v1/image"
    # -H "accept: application/json"
    # -H "Authorization: DEADBEEF"
    # -H "content-type: multipart/form-data"
    # -F "file=@linux;type="
    # -F "q={ "description": "Example image",
    #         "type": "Kernel",
    #         "public": false,
    #         "known_good": true } "
    headers = {'Authorization': module.params['token']}
    url = urljoin(module.params['url'], "/api/v1/image")
    files = {'file': open(module.params['path'], 'rb')}
    data = {'q': json.dumps({
                 'description': module.params['description'],
                 'type': module.params['type'],
                 'arch': module.params['arch'],
                 'known_good': module.params['known_good'],
                 'public': module.params['public'],
             })
           }
    r = requests.post(url, files=files, data=data, headers=headers)
    if r.status_code != 201:
        msg = ("Error fetching {}, HTTP {} {}\nrequest data: {}\nresult json: {}".
                format(url, r.status_code, r.reason, data, r.json()))
        module.fail_json(msg=msg, **result)
    result['json'] = r.json()
    result['changed'] = True

    # in the event of a successful module execution, you will want to
    # simple AnsibleModule.exit_json(), passing the key/value results
    module.exit_json(**result)

def main():
    run_module()

if __name__ == '__main__':
    main()
